File: src/protobuf/java/src/main/java/com/google/protobuf/MessageLite.java
// protocol buffers - google's data interchange format
// copyright 2008 google inc.  all rights reserved.
// http://code.google.com/p/protobuf/
//
// redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * neither the name of google inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// this software is provided by the copyright holders and contributors
// "as is" and any express or implied warranties, including, but not
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed. in no event shall the copyright
// owner or contributors be liable for any direct, indirect, incidental,
// special, exemplary, or consequential damages (including, but not
// limited to, procurement of substitute goods or services; loss of use,
// data, or profits; or business interruption) however caused and on any
// theory of liability, whether in contract, strict liability, or tort
// (including negligence or otherwise) arising in any way out of the use
// of this software, even if advised of the possibility of such damage.

// todo(kenton):  use generics?  e.g. builder<buildertype extends builder>, then
//   mergefrom*() could return buildertype for better type-safety.

package com.google.protobuf;

import java.io.ioexception;
import java.io.inputstream;
import java.io.outputstream;

/**
 * abstract interface implemented by protocol message objects.
 *
 * <p>this interface is implemented by all protocol message objects.  non-lite
 * messages additionally implement the message interface, which is a subclass
 * of messagelite.  use messagelite instead when you only need the subset of
 * features which it supports -- namely, nothing that uses descriptors or
 * reflection.  you can instruct the protocol compiler to generate classes
 * which implement only messagelite, not the full message interface, by adding
 * the follow line to the .proto file:
 * <pre>
 *   option optimize_for = lite_runtime;
 * </pre>
 *
 * <p>this is particularly useful on resource-constrained systems where the
 * full protocol buffers runtime library is too big.
 *
 * <p>note that on non-constrained systems (e.g. servers) when you need to link
 * in lots of protocol definitions, a better way to reduce total code footprint
 * is to use {@code optimize_for = code_size}.  this will make the generated
 * code smaller while still supporting all the same features (at the expense of
 * speed).  {@code optimize_for = lite_runtime} is best when you only have a
 * small number of message types linked into your binary, in which case the
 * size of the protocol buffers runtime itself is the biggest problem.
 *
 * @author kenton@google.com kenton varda
 */
public interface messagelite extends messageliteorbuilder {


  /**
   * serializes the message and writes it to {@code output}.  this does not
   * flush or close the stream.
   */
  void writeto(codedoutputstream output) throws ioexception;

  /**
   * get the number of bytes required to encode this message.  the result
   * is only computed on the first call and memoized after that.
   */
  int getserializedsize();


  /**
   * gets the parser for a message of the same type as this message.
   */
  parser<? extends messagelite> getparserfortype();

  // -----------------------------------------------------------------
  // convenience methods.

  /**
   * serializes the message to a {@code bytestring} and returns it. this is
   * just a trivial wrapper around
   * {@link #writeto(codedoutputstream)}.
   */
  bytestring tobytestring();

  /**
   * serializes the message to a {@code byte} array and returns it.  this is
   * just a trivial wrapper around
   * {@link #writeto(codedoutputstream)}.
   */
  byte[] tobytearray();

  /**
   * serializes the message and writes it to {@code output}.  this is just a
   * trivial wrapper around {@link #writeto(codedoutputstream)}.  this does
   * not flush or close the stream.
   * <p>
   * note:  protocol buffers are not self-delimiting.  therefore, if you write
   * any more data to the stream after the message, you must somehow ensure
   * that the parser on the receiving end does not interpret this as being
   * part of the protocol message.  this can be done e.g. by writing the size
   * of the message before the data, then making sure to limit the input to
   * that size on the receiving end (e.g. by wrapping the inputstream in one
   * which limits the input).  alternatively, just use
   * {@link #writedelimitedto(outputstream)}.
   */
  void writeto(outputstream output) throws ioexception;

  /**
   * like {@link #writeto(outputstream)}, but writes the size of the message
   * as a varint before writing the data.  this allows more data to be written
   * to the stream after the message without the need to delimit the message
   * data yourself.  use {@link builder#mergedelimitedfrom(inputstream)} (or
   * the static method {@code yourmessagetype.parsedelimitedfrom(inputstream)})
   * to parse messages written by this method.
   */
  void writedelimitedto(outputstream output) throws ioexception;

  // =================================================================
  // builders

  /**
   * constructs a new builder for a message of the same type as this message.
   */
  builder newbuilderfortype();

  /**
   * constructs a builder initialized with the current message.  use this to
   * derive a new message from the current one.
   */
  builder tobuilder();

  /**
   * abstract interface implemented by protocol message builders.
   */
  interface builder extends messageliteorbuilder, cloneable {
    /** resets all fields to their default values. */
    builder clear();

    /**
     * constructs the message based on the state of the builder. subsequent
     * changes to the builder will not affect the returned message.
     * @throws uninitializedmessageexception the message is missing one or more
     *         required fields (i.e. {@link #isinitialized()} returns false).
     *         use {@link #buildpartial()} to bypass this check.
     */
    messagelite build();

    /**
     * like {@link #build()}, but does not throw an exception if the message
     * is missing required fields.  instead, a partial message is returned.
     * subsequent changes to the builder will not affect the returned message.
     */
    messagelite buildpartial();

    /**
     * clones the builder.
     * @see object#clone()
     */
    builder clone();

    /**
     * parses a message of this type from the input and merges it with this
     * message.
     *
     * <p>warning:  this does not verify that all required fields are present in
     * the input message.  if you call {@link #build()} without setting all
     * required fields, it will throw an {@link uninitializedmessageexception},
     * which is a {@code runtimeexception} and thus might not be caught.  there
     * are a few good ways to deal with this:
     * <ul>
     *   <li>call {@link #isinitialized()} to verify that all required fields
     *       are set before building.
     *   <li>use {@code buildpartial()} to build, which ignores missing
     *       required fields.
     * </ul>
     *
     * <p>note:  the caller should call
     * {@link codedinputstream#checklasttagwas(int)} after calling this to
     * verify that the last tag seen was the appropriate end-group tag,
     * or zero for eof.
     */
    builder mergefrom(codedinputstream input) throws ioexception;

    /**
     * like {@link builder#mergefrom(codedinputstream)}, but also
     * parses extensions.  the extensions that you want to be able to parse
     * must be registered in {@code extensionregistry}.  extensions not in
     * the registry will be treated as unknown fields.
     */
    builder mergefrom(codedinputstream input,
                      extensionregistrylite extensionregistry)
                      throws ioexception;

    // ---------------------------------------------------------------
    // convenience methods.

    /**
     * parse {@code data} as a message of this type and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream)}.
     *
     * @return this
     */
    builder mergefrom(bytestring data) throws invalidprotocolbufferexception;

    /**
     * parse {@code data} as a message of this type and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream,extensionregistrylite)}.
     *
     * @return this
     */
    builder mergefrom(bytestring data,
                      extensionregistrylite extensionregistry)
                      throws invalidprotocolbufferexception;

    /**
     * parse {@code data} as a message of this type and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream)}.
     *
     * @return this
     */
    builder mergefrom(byte[] data) throws invalidprotocolbufferexception;

    /**
     * parse {@code data} as a message of this type and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream)}.
     *
     * @return this
     */
    builder mergefrom(byte[] data, int off, int len)
                      throws invalidprotocolbufferexception;

    /**
     * parse {@code data} as a message of this type and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream,extensionregistrylite)}.
     *
     * @return this
     */
    builder mergefrom(byte[] data,
                      extensionregistrylite extensionregistry)
                      throws invalidprotocolbufferexception;

    /**
     * parse {@code data} as a message of this type and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream,extensionregistrylite)}.
     *
     * @return this
     */
    builder mergefrom(byte[] data, int off, int len,
                      extensionregistrylite extensionregistry)
                      throws invalidprotocolbufferexception;

    /**
     * parse a message of this type from {@code input} and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream)}.  note that this method always
     * reads the <i>entire</i> input (unless it throws an exception).  if you
     * want it to stop earlier, you will need to wrap your input in some
     * wrapper stream that limits reading.  or, use
     * {@link messagelite#writedelimitedto(outputstream)} to write your message
     * and {@link #mergedelimitedfrom(inputstream)} to read it.
     * <p>
     * despite usually reading the entire input, this does not close the stream.
     *
     * @return this
     */
    builder mergefrom(inputstream input) throws ioexception;

    /**
     * parse a message of this type from {@code input} and merge it with the
     * message being built.  this is just a small wrapper around
     * {@link #mergefrom(codedinputstream,extensionregistrylite)}.
     *
     * @return this
     */
    builder mergefrom(inputstream input,
                      extensionregistrylite extensionregistry)
                      throws ioexception;

    /**
     * like {@link #mergefrom(inputstream)}, but does not read until eof.
     * instead, the size of the message (encoded as a varint) is read first,
     * then the message data.  use
     * {@link messagelite#writedelimitedto(outputstream)} to write messages in
     * this format.
     *
     * @return true if successful, or false if the stream is at eof when the
     *         method starts.  any other error (including reaching eof during
     *         parsing) will cause an exception to be thrown.
     */
    boolean mergedelimitedfrom(inputstream input)
                               throws ioexception;

    /**
     * like {@link #mergedelimitedfrom(inputstream)} but supporting extensions.
     */
    boolean mergedelimitedfrom(inputstream input,
                               extensionregistrylite extensionregistry)
                               throws ioexception;
  }
}

File: src/protobuf/src/google/protobuf/message_lite.cc
// protocol buffers - google's data interchange format
// copyright 2008 google inc.  all rights reserved.
// http://code.google.com/p/protobuf/
//
// redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * neither the name of google inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// this software is provided by the copyright holders and contributors
// "as is" and any express or implied warranties, including, but not
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed. in no event shall the copyright
// owner or contributors be liable for any direct, indirect, incidental,
// special, exemplary, or consequential damages (including, but not
// limited to, procurement of substitute goods or services; loss of use,
// data, or profits; or business interruption) however caused and on any
// theory of liability, whether in contract, strict liability, or tort
// (including negligence or otherwise) arising in any way out of the use
// of this software, even if advised of the possibility of such damage.

// authors: wink@google.com (wink saville),
//          kenton@google.com (kenton varda)
//  based on original protocol buffers design by
//  sanjay ghemawat, jeff dean, and others.

#include <google/protobuf/message_lite.h>
#include <string>
#include <google/protobuf/stubs/common.h>
#include <google/protobuf/io/coded_stream.h>
#include <google/protobuf/io/zero_copy_stream_impl_lite.h>
#include <google/protobuf/stubs/stl_util.h>

namespace google {
namespace protobuf {

messagelite::~messagelite() {}

string messagelite::initializationerrorstring() const {
  return "(cannot determine missing fields for lite message)";
}

namespace {

// when serializing, we first compute the byte size, then serialize the message.
// if serialization produces a different number of bytes than expected, we
// call this function, which crashes.  the problem could be due to a bug in the
// protobuf implementation but is more likely caused by concurrent modification
// of the message.  this function attempts to distinguish between the two and
// provide a useful error message.
void bytesizeconsistencyerror(int byte_size_before_serialization,
                              int byte_size_after_serialization,
                              int bytes_produced_by_serialization) {
  google_check_eq(byte_size_before_serialization, byte_size_after_serialization)
      << "protocol message was modified concurrently during serialization.";
  google_check_eq(bytes_produced_by_serialization, byte_size_before_serialization)
      << "byte size calculation and serialization were inconsistent.  this "
         "may indicate a bug in protocol buffers or it may be caused by "
         "concurrent modification of the message.";
  google_log(fatal) << "this shouldn't be called if all the sizes are equal.";
}

string initializationerrormessage(const char* action,
                                  const messagelite& message) {
  // note:  we want to avoid depending on strutil in the lite library, otherwise
  //   we'd use:
  //
  // return strings::substitute(
  //   "can't $0 message of type \"$1\" because it is missing required "
  //   "fields: $2",
  //   action, message.gettypename(),
  //   message.initializationerrorstring());

  string result;
  result += "can't ";
  result += action;
  result += " message of type \"";
  result += message.gettypename();
  result += "\" because it is missing required fields: ";
  result += message.initializationerrorstring();
  return result;
}

// several of the parse methods below just do one thing and then call another
// method.  in a naive implementation, we might have parsefromstring() call
// parsefromarray() which would call parsefromzerocopystream() which would call
// parsefromcodedstream() which would call mergefromcodedstream() which would
// call mergepartialfromcodedstream().  however, when parsing very small
// messages, every function call introduces significant overhead.  to avoid
// this without reproducing code, we use these forced-inline helpers.
//
// note:  gcc only allows google_attribute_always_inline on declarations, not
//   definitions.
inline bool inlinemergefromcodedstream(io::codedinputstream* input,
                                       messagelite* message)
                                       google_attribute_always_inline;
inline bool inlineparsefromcodedstream(io::codedinputstream* input,
                                       messagelite* message)
                                       google_attribute_always_inline;
inline bool inlineparsepartialfromcodedstream(io::codedinputstream* input,
                                              messagelite* message)
                                              google_attribute_always_inline;
inline bool inlineparsefromarray(const void* data, int size,
                                 messagelite* message)
                                 google_attribute_always_inline;
inline bool inlineparsepartialfromarray(const void* data, int size,
                                        messagelite* message)
                                        google_attribute_always_inline;

bool inlinemergefromcodedstream(io::codedinputstream* input,
                                messagelite* message) {
  if (!message->mergepartialfromcodedstream(input)) return false;
  if (!message->isinitialized()) {
    google_log(error) << initializationerrormessage("parse", *message);
    return false;
  }
  return true;
}

bool inlineparsefromcodedstream(io::codedinputstream* input,
                                messagelite* message) {
  message->clear();
  return inlinemergefromcodedstream(input, message);
}

bool inlineparsepartialfromcodedstream(io::codedinputstream* input,
                                       messagelite* message) {
  message->clear();
  return message->mergepartialfromcodedstream(input);
}

bool inlineparsefromarray(const void* data, int size, messagelite* message) {
  io::codedinputstream input(reinterpret_cast<const uint8*>(data), size);
  return inlineparsefromcodedstream(&input, message) &&
         input.consumedentiremessage();
}

bool inlineparsepartialfromarray(const void* data, int size,
                                 messagelite* message) {
  io::codedinputstream input(reinterpret_cast<const uint8*>(data), size);
  return inlineparsepartialfromcodedstream(&input, message) &&
         input.consumedentiremessage();
}

}  // namespace

bool messagelite::mergefromcodedstream(io::codedinputstream* input) {
  return inlinemergefromcodedstream(input, this);
}

bool messagelite::parsefromcodedstream(io::codedinputstream* input) {
  return inlineparsefromcodedstream(input, this);
}

bool messagelite::parsepartialfromcodedstream(io::codedinputstream* input) {
  return inlineparsepartialfromcodedstream(input, this);
}

bool messagelite::parsefromzerocopystream(io::zerocopyinputstream* input) {
  io::codedinputstream decoder(input);
  return parsefromcodedstream(&decoder) && decoder.consumedentiremessage();
}

bool messagelite::parsepartialfromzerocopystream(
    io::zerocopyinputstream* input) {
  io::codedinputstream decoder(input);
  return parsepartialfromcodedstream(&decoder) &&
         decoder.consumedentiremessage();
}

bool messagelite::parsefromboundedzerocopystream(
    io::zerocopyinputstream* input, int size) {
  io::codedinputstream decoder(input);
  decoder.pushlimit(size);
  return parsefromcodedstream(&decoder) &&
         decoder.consumedentiremessage() &&
         decoder.bytesuntillimit() == 0;
}

bool messagelite::parsepartialfromboundedzerocopystream(
    io::zerocopyinputstream* input, int size) {
  io::codedinputstream decoder(input);
  decoder.pushlimit(size);
  return parsepartialfromcodedstream(&decoder) &&
         decoder.consumedentiremessage() &&
         decoder.bytesuntillimit() == 0;
}

bool messagelite::parsefromstring(const string& data) {
  return inlineparsefromarray(data.data(), data.size(), this);
}

bool messagelite::parsepartialfromstring(const string& data) {
  return inlineparsepartialfromarray(data.data(), data.size(), this);
}

bool messagelite::parsefromarray(const void* data, int size) {
  return inlineparsefromarray(data, size, this);
}

bool messagelite::parsepartialfromarray(const void* data, int size) {
  return inlineparsepartialfromarray(data, size, this);
}


// ===================================================================

uint8* messagelite::serializewithcachedsizestoarray(uint8* target) const {
  // we only optimize this when using optimize_for = speed.  in other cases
  // we just use the codedoutputstream path.
  int size = getcachedsize();
  io::arrayoutputstream out(target, size);
  io::codedoutputstream coded_out(&out);
  serializewithcachedsizes(&coded_out);
  google_check(!coded_out.haderror());
  return target + size;
}

bool messagelite::serializetocodedstream(io::codedoutputstream* output) const {
  google_dcheck(isinitialized()) << initializationerrormessage("serialize", *this);
  return serializepartialtocodedstream(output);
}

bool messagelite::serializepartialtocodedstream(
    io::codedoutputstream* output) const {
  const int size = bytesize();  // force size to be cached.
  uint8* buffer = output->getdirectbufferfornbytesandadvance(size);
  if (buffer != null) {
    uint8* end = serializewithcachedsizestoarray(buffer);
    if (end - buffer != size) {
      bytesizeconsistencyerror(size, bytesize(), end - buffer);
    }
    return true;
  } else {
    int original_byte_count = output->bytecount();
    serializewithcachedsizes(output);
    if (output->haderror()) {
      return false;
    }
    int final_byte_count = output->bytecount();

    if (final_byte_count - original_byte_count != size) {
      bytesizeconsistencyerror(size, bytesize(),
                               final_byte_count - original_byte_count);
    }

    return true;
  }
}

bool messagelite::serializetozerocopystream(
    io::zerocopyoutputstream* output) const {
  io::codedoutputstream encoder(output);
  return serializetocodedstream(&encoder);
}

bool messagelite::serializepartialtozerocopystream(
    io::zerocopyoutputstream* output) const {
  io::codedoutputstream encoder(output);
  return serializepartialtocodedstream(&encoder);
}

bool messagelite::appendtostring(string* output) const {
  google_dcheck(isinitialized()) << initializationerrormessage("serialize", *this);
  return appendpartialtostring(output);
}

bool messagelite::appendpartialtostring(string* output) const {
  int old_size = output->size();
  int byte_size = bytesize();
  stlstringresizeuninitialized(output, old_size + byte_size);
  uint8* start = reinterpret_cast<uint8*>(string_as_array(output) + old_size);
  uint8* end = serializewithcachedsizestoarray(start);
  if (end - start != byte_size) {
    bytesizeconsistencyerror(byte_size, bytesize(), end - start);
  }
  return true;
}

bool messagelite::serializetostring(string* output) const {
  output->clear();
  return appendtostring(output);
}

bool messagelite::serializepartialtostring(string* output) const {
  output->clear();
  return appendpartialtostring(output);
}

bool messagelite::serializetoarray(void* data, int size) const {
  google_dcheck(isinitialized()) << initializationerrormessage("serialize", *this);
  return serializepartialtoarray(data, size);
}

bool messagelite::serializepartialtoarray(void* data, int size) const {
  int byte_size = bytesize();
  if (size < byte_size) return false;
  uint8* start = reinterpret_cast<uint8*>(data);
  uint8* end = serializewithcachedsizestoarray(start);
  if (end - start != byte_size) {
    bytesizeconsistencyerror(byte_size, bytesize(), end - start);
  }
  return true;
}

string messagelite::serializeasstring() const {
  // if the compiler implements the (named) return value optimization,
  // the local variable 'result' will not actually reside on the stack
  // of this function, but will be overlaid with the object that the
  // caller supplied for the return value to be constructed in.
  string output;
  if (!appendtostring(&output))
    output.clear();
  return output;
}

string messagelite::serializepartialasstring() const {
  string output;
  if (!appendpartialtostring(&output))
    output.clear();
  return output;
}

}  // namespace protobuf
}  // namespace google
